Index: src/main/java/com/sun/el/parser/SimpleNode.java
===================================================================
--- src/main/java/com/sun/el/parser/SimpleNode.java	(revision 352)
+++ src/main/java/com/sun/el/parser/SimpleNode.java	(working copy)
@@ -62,6 +62,9 @@
 
     protected String image;
 
+    protected Token firstToken;
+    protected Token lastToken;
+
     public SimpleNode(int i) {
         id = i;
     }
@@ -99,6 +102,53 @@
         return (children == null) ? 0 : children.length;
     }
 
+    public Token jjtGetFirstToken() {
+        return firstToken;
+    }
+
+    public void jjtSetFirstToken(Token token) {
+        this.firstToken = token;
+    }
+
+    public Token jjtGetLastToken() {
+        return lastToken;
+    }
+
+    public void jjtSetLastToken(Token token) {
+        this.lastToken = token;
+    }
+
+    public int startOffset() {
+        return jjtGetFirstToken().offset;
+    }
+
+    public int endOffset() {
+        //why this happens???
+        if (jjtGetLastToken().image == null) {
+            System.err.println("ERROR: lastToken image is null! : " + jjtGetLastToken());
+            return jjtGetLastToken().offset;
+        } else {
+            //#181357 - At the end of the tokens sequence there is the EOF token (kind==0) with
+            //zero lenght but WRONG start offset - it points at the previous offset (real file end - 1).
+            //so tokens of simple code may look like:
+            //Token( 0; 'h1')
+            //Token( 2; ' ')
+            //Token( 3; '{')
+            //Token( 4; ' ')
+            //Token( 5; '}')
+            //Token( 5; '') <-- here the (EOF) token should apparently have offset set to 6.
+            //
+            //sometimes there are even two EOF tokens at the end of the sequence with the same wrong offset!
+            //
+            //why this happens is a mystery to me, maybe caused by some changes to the default javacc
+            //lexing, but I am not sure so I'll workaround it here.
+            if(jjtGetLastToken().kind == ELParserConstants.EOF) {
+                return jjtGetLastToken().offset + 1;
+            } else {
+                return jjtGetLastToken().offset + jjtGetLastToken().image.length();
+            }
+        }
+    }
     /*
      * You can override these two methods in subclasses of SimpleNode to
      * customize the way the node appears when the tree is dumped. If your
Index: src/main/java/com/sun/el/parser/PatchedELParserTokenManager.java
===================================================================
--- src/main/java/com/sun/el/parser/PatchedELParserTokenManager.java	(revision 0)
+++ src/main/java/com/sun/el/parser/PatchedELParserTokenManager.java	(revision 0)
@@ -0,0 +1,28 @@
+/*
+ * To change this template, choose Tools | Templates
+ * and open the template in the editor.
+ */
+package com.sun.el.parser;
+
+/**
+ *
+ * @author Martin Fousek <marfous@netbeans.org>
+ */
+public class PatchedELParserTokenManager extends ELParserTokenManager
+{
+  public PatchedELParserTokenManager(SimpleCharStream stream, int lexState)
+  {
+    super(stream, lexState);
+  }
+
+  public PatchedELParserTokenManager(SimpleCharStream stream) {
+    super(stream);
+  }
+
+  protected Token jjFillToken()
+  {
+    Token t = super.jjFillToken();
+    t.offset = this.input_stream.offset();
+    return t;
+  }
+}
\ No newline at end of file
Index: src/main/java/com/sun/el/parser/ELParser.java
===================================================================
--- src/main/java/com/sun/el/parser/ELParser.java	(revision 352)
+++ src/main/java/com/sun/el/parser/ELParser.java	(working copy)
@@ -43,10 +43,13 @@
 import java.io.StringReader;
 import javax.el.ELException;
 public class ELParser/*@bgen(jjtree)*/implements ELParserTreeConstants, ELParserConstants {/*@bgen(jjtree)*/
-  protected JJTELParserState jjtree = new JJTELParserState();public static Node parse(String ref) throws ELException
+  protected JJTELParserState jjtree = new JJTELParserState();
+    public static Node parse(String ref) throws ELException
     {
         try {
-                return (new ELParser(new StringReader(ref))).CompositeExpression();
+            ELParser parser = new ELParser(new StringReader(ref));
+            parser.token_source = new PatchedELParserTokenManager(parser.jj_input_stream);
+            return parser.CompositeExpression();
         } catch (ParseException pe) {
                 throw new ELException(pe.getMessage());
         }
@@ -62,6 +65,8 @@
   AstCompositeExpression jjtn000 = new AstCompositeExpression(JJTCOMPOSITEEXPRESSION);
   boolean jjtc000 = true;
   jjtree.openNodeScope(jjtn000);
+  jjtn000.jjtSetFirstToken(getToken(1));
+  Token t = null;
     try {
       label_1:
       while (true) {
@@ -92,9 +97,10 @@
         }
       }
       jj_consume_token(0);
-                                                                                    jjtree.closeNodeScope(jjtn000, true);
-                                                                                    jjtc000 = false;
-                                                                                    {if (true) return jjtn000;}
+        jjtree.closeNodeScope(jjtn000, true);
+        jjtc000 = false;
+        jjtn000.jjtSetLastToken(getToken(0));
+        {if (true) return jjtn000;}
     } catch (Throwable jjte000) {
           if (jjtc000) {
             jjtree.clearNodeScope(jjtn000);
@@ -112,6 +118,7 @@
     } finally {
           if (jjtc000) {
             jjtree.closeNodeScope(jjtn000, true);
+            jjtn000.jjtSetLastToken(getToken(0));
           }
     }
     throw new Error("Missing return statement in function");
@@ -125,15 +132,18 @@
                                                /*@bgen(jjtree) LiteralExpression */
                                                 AstLiteralExpression jjtn000 = new AstLiteralExpression(JJTLITERALEXPRESSION);
                                                 boolean jjtc000 = true;
-                                                jjtree.openNodeScope(jjtn000);Token t = null;
+                                                jjtree.openNodeScope(jjtn000);
+                                                jjtn000.jjtSetFirstToken(getToken(1));Token t = null;
     try {
       t = jj_consume_token(LITERAL_EXPRESSION);
                                  jjtree.closeNodeScope(jjtn000, true);
                                  jjtc000 = false;
+                                 jjtn000.jjtSetLastToken(getToken(0));
                                  jjtn000.setImage(t.image);
     } finally {
           if (jjtc000) {
             jjtree.closeNodeScope(jjtn000, true);
+            jjtn000.jjtSetLastToken(getToken(0));
           }
     }
   }
@@ -147,6 +157,7 @@
   AstDeferredExpression jjtn000 = new AstDeferredExpression(JJTDEFERREDEXPRESSION);
   boolean jjtc000 = true;
   jjtree.openNodeScope(jjtn000);
+  jjtn000.jjtSetFirstToken(getToken(1));
     try {
       jj_consume_token(START_DEFERRED_EXPRESSION);
       Expression();
@@ -168,6 +179,7 @@
     } finally {
           if (jjtc000) {
             jjtree.closeNodeScope(jjtn000, true);
+            jjtn000.jjtSetLastToken(getToken(0));
           }
     }
   }
@@ -181,6 +193,7 @@
   AstDynamicExpression jjtn000 = new AstDynamicExpression(JJTDYNAMICEXPRESSION);
   boolean jjtc000 = true;
   jjtree.openNodeScope(jjtn000);
+  jjtn000.jjtSetFirstToken(getToken(1));
     try {
       jj_consume_token(START_DYNAMIC_EXPRESSION);
       Expression();
@@ -202,6 +215,7 @@
     } finally {
           if (jjtc000) {
             jjtree.closeNodeScope(jjtn000, true);
+            jjtn000.jjtSetLastToken(getToken(0));
           }
     }
   }
@@ -228,6 +242,7 @@
                                                 AstChoice jjtn001 = new AstChoice(JJTCHOICE);
                                                 boolean jjtc001 = true;
                                                 jjtree.openNodeScope(jjtn001);
+                                                jjtn001.jjtSetFirstToken(getToken(1));
       try {
         Choice();
       } catch (Throwable jjte001) {
@@ -247,6 +262,7 @@
       } finally {
                                                 if (jjtc001) {
                                                   jjtree.closeNodeScope(jjtn001,  3);
+                                                  jjtn001.jjtSetLastToken(getToken(0));
                                                 }
       }
       break;
@@ -288,6 +304,7 @@
                                AstOr jjtn001 = new AstOr(JJTOR);
                                boolean jjtc001 = true;
                                jjtree.openNodeScope(jjtn001);
+                               jjtn001.jjtSetFirstToken(getToken(1));
       try {
         And();
       } catch (Throwable jjte001) {
@@ -307,6 +324,7 @@
       } finally {
                                if (jjtc001) {
                                  jjtree.closeNodeScope(jjtn001,  2);
+                                 jjtn001.jjtSetLastToken(getToken(0));
                                }
       }
     }
@@ -344,6 +362,7 @@
                                       AstAnd jjtn001 = new AstAnd(JJTAND);
                                       boolean jjtc001 = true;
                                       jjtree.openNodeScope(jjtn001);
+                                      jjtn001.jjtSetFirstToken(getToken(1));
       try {
         Equality();
       } catch (Throwable jjte001) {
@@ -363,6 +382,7 @@
       } finally {
                                       if (jjtc001) {
                                         jjtree.closeNodeScope(jjtn001,  2);
+                                        jjtn001.jjtSetLastToken(getToken(0));
                                       }
       }
     }
@@ -405,6 +425,7 @@
                                  AstEqual jjtn001 = new AstEqual(JJTEQUAL);
                                  boolean jjtc001 = true;
                                  jjtree.openNodeScope(jjtn001);
+                                 jjtn001.jjtSetFirstToken(getToken(1));
         try {
           Compare();
         } catch (Throwable jjte001) {
@@ -424,6 +445,7 @@
         } finally {
                                  if (jjtc001) {
                                    jjtree.closeNodeScope(jjtn001,  2);
+                                   jjtn001.jjtSetLastToken(getToken(0));
                                  }
         }
         break;
@@ -444,6 +466,7 @@
                                  AstNotEqual jjtn002 = new AstNotEqual(JJTNOTEQUAL);
                                  boolean jjtc002 = true;
                                  jjtree.openNodeScope(jjtn002);
+                                 jjtn002.jjtSetFirstToken(getToken(1));
         try {
           Compare();
         } catch (Throwable jjte002) {
@@ -463,6 +486,7 @@
         } finally {
                                  if (jjtc002) {
                                    jjtree.closeNodeScope(jjtn002,  2);
+                                   jjtn002.jjtSetLastToken(getToken(0));
                                  }
         }
         break;
@@ -515,6 +539,7 @@
                                  AstLessThan jjtn001 = new AstLessThan(JJTLESSTHAN);
                                  boolean jjtc001 = true;
                                  jjtree.openNodeScope(jjtn001);
+                                 jjtn001.jjtSetFirstToken(getToken(1));
         try {
           Math();
         } catch (Throwable jjte001) {
@@ -534,6 +559,7 @@
         } finally {
                                  if (jjtc001) {
                                    jjtree.closeNodeScope(jjtn001,  2);
+                                   jjtn001.jjtSetLastToken(getToken(0));
                                  }
         }
         break;
@@ -554,6 +580,7 @@
                                  AstGreaterThan jjtn002 = new AstGreaterThan(JJTGREATERTHAN);
                                  boolean jjtc002 = true;
                                  jjtree.openNodeScope(jjtn002);
+                                 jjtn002.jjtSetFirstToken(getToken(1));
         try {
           Math();
         } catch (Throwable jjte002) {
@@ -573,6 +600,7 @@
         } finally {
                                  if (jjtc002) {
                                    jjtree.closeNodeScope(jjtn002,  2);
+                                   jjtn002.jjtSetLastToken(getToken(0));
                                  }
         }
         break;
@@ -593,6 +621,7 @@
                                  AstLessThanEqual jjtn003 = new AstLessThanEqual(JJTLESSTHANEQUAL);
                                  boolean jjtc003 = true;
                                  jjtree.openNodeScope(jjtn003);
+                                 jjtn003.jjtSetFirstToken(getToken(1));
         try {
           Math();
         } catch (Throwable jjte003) {
@@ -612,6 +641,7 @@
         } finally {
                                  if (jjtc003) {
                                    jjtree.closeNodeScope(jjtn003,  2);
+                                   jjtn003.jjtSetLastToken(getToken(0));
                                  }
         }
         break;
@@ -632,6 +662,7 @@
                                  AstGreaterThanEqual jjtn004 = new AstGreaterThanEqual(JJTGREATERTHANEQUAL);
                                  boolean jjtc004 = true;
                                  jjtree.openNodeScope(jjtn004);
+                                 jjtn004.jjtSetFirstToken(getToken(1));
         try {
           Math();
         } catch (Throwable jjte004) {
@@ -651,6 +682,7 @@
         } finally {
                                  if (jjtc004) {
                                    jjtree.closeNodeScope(jjtn004,  2);
+                                   jjtn004.jjtSetLastToken(getToken(0));
                                  }
         }
         break;
@@ -685,6 +717,7 @@
                           AstPlus jjtn001 = new AstPlus(JJTPLUS);
                           boolean jjtc001 = true;
                           jjtree.openNodeScope(jjtn001);
+                          jjtn001.jjtSetFirstToken(getToken(1));
         try {
           Multiplication();
         } catch (Throwable jjte001) {
@@ -704,6 +737,7 @@
         } finally {
                           if (jjtc001) {
                             jjtree.closeNodeScope(jjtn001,  2);
+                            jjtn001.jjtSetLastToken(getToken(0));
                           }
         }
         break;
@@ -712,6 +746,7 @@
                            AstMinus jjtn002 = new AstMinus(JJTMINUS);
                            boolean jjtc002 = true;
                            jjtree.openNodeScope(jjtn002);
+                           jjtn002.jjtSetFirstToken(getToken(1));
         try {
           Multiplication();
         } catch (Throwable jjte002) {
@@ -731,6 +766,7 @@
         } finally {
                            if (jjtc002) {
                              jjtree.closeNodeScope(jjtn002,  2);
+                             jjtn002.jjtSetLastToken(getToken(0));
                            }
         }
         break;
@@ -768,6 +804,7 @@
                           AstMult jjtn001 = new AstMult(JJTMULT);
                           boolean jjtc001 = true;
                           jjtree.openNodeScope(jjtn001);
+                          jjtn001.jjtSetFirstToken(getToken(1));
         try {
           Unary();
         } catch (Throwable jjte001) {
@@ -787,6 +824,7 @@
         } finally {
                           if (jjtc001) {
                             jjtree.closeNodeScope(jjtn001,  2);
+                            jjtn001.jjtSetLastToken(getToken(0));
                           }
         }
         break;
@@ -807,6 +845,7 @@
                                    AstDiv jjtn002 = new AstDiv(JJTDIV);
                                    boolean jjtc002 = true;
                                    jjtree.openNodeScope(jjtn002);
+                                   jjtn002.jjtSetFirstToken(getToken(1));
         try {
           Unary();
         } catch (Throwable jjte002) {
@@ -826,6 +865,7 @@
         } finally {
                                    if (jjtc002) {
                                      jjtree.closeNodeScope(jjtn002,  2);
+                                     jjtn002.jjtSetLastToken(getToken(0));
                                    }
         }
         break;
@@ -846,6 +886,7 @@
                                    AstMod jjtn003 = new AstMod(JJTMOD);
                                    boolean jjtc003 = true;
                                    jjtree.openNodeScope(jjtn003);
+                                   jjtn003.jjtSetFirstToken(getToken(1));
         try {
           Unary();
         } catch (Throwable jjte003) {
@@ -865,6 +906,7 @@
         } finally {
                                    if (jjtc003) {
                                      jjtree.closeNodeScope(jjtn003,  2);
+                                     jjtn003.jjtSetLastToken(getToken(0));
                                    }
         }
         break;
@@ -887,6 +929,7 @@
                           AstNegative jjtn001 = new AstNegative(JJTNEGATIVE);
                           boolean jjtc001 = true;
                           jjtree.openNodeScope(jjtn001);
+                          jjtn001.jjtSetFirstToken(getToken(1));
       try {
         Unary();
       } catch (Throwable jjte001) {
@@ -906,6 +949,7 @@
       } finally {
                           if (jjtc001) {
                             jjtree.closeNodeScope(jjtn001, true);
+                            jjtn001.jjtSetLastToken(getToken(0));
                           }
       }
       break;
@@ -926,6 +970,7 @@
                                   AstNot jjtn002 = new AstNot(JJTNOT);
                                   boolean jjtc002 = true;
                                   jjtree.openNodeScope(jjtn002);
+                                  jjtn002.jjtSetFirstToken(getToken(1));
       try {
         Unary();
       } catch (Throwable jjte002) {
@@ -945,6 +990,7 @@
       } finally {
                                   if (jjtc002) {
                                     jjtree.closeNodeScope(jjtn002, true);
+                                    jjtn002.jjtSetLastToken(getToken(0));
                                   }
       }
       break;
@@ -953,6 +999,7 @@
                           AstEmpty jjtn003 = new AstEmpty(JJTEMPTY);
                           boolean jjtc003 = true;
                           jjtree.openNodeScope(jjtn003);
+                          jjtn003.jjtSetFirstToken(getToken(1));
       try {
         Unary();
       } catch (Throwable jjte003) {
@@ -972,6 +1019,7 @@
       } finally {
                           if (jjtc003) {
                             jjtree.closeNodeScope(jjtn003, true);
+                            jjtn003.jjtSetLastToken(getToken(0));
                           }
       }
       break;
@@ -1000,6 +1048,7 @@
           AstValue jjtn001 = new AstValue(JJTVALUE);
           boolean jjtc001 = true;
           jjtree.openNodeScope(jjtn001);
+          jjtn001.jjtSetFirstToken(getToken(1));
     try {
       ValuePrefix();
       label_8:
@@ -1032,6 +1081,7 @@
     } finally {
           if (jjtc001) {
             jjtree.closeNodeScope(jjtn001, jjtree.nodeArity() > 1);
+            jjtn001.jjtSetLastToken(getToken(0));
           }
     }
   }
@@ -1089,6 +1139,7 @@
                                 AstDotSuffix jjtn000 = new AstDotSuffix(JJTDOTSUFFIX);
                                 boolean jjtc000 = true;
                                 jjtree.openNodeScope(jjtn000);Token t = null;
+                                jjtn000.jjtSetFirstToken(getToken(2));
     try {
       jj_consume_token(DOT);
       t = jj_consume_token(IDENTIFIER);
@@ -1098,6 +1149,7 @@
             AstMethodArguments jjtn001 = new AstMethodArguments(JJTMETHODARGUMENTS);
             boolean jjtc001 = true;
             jjtree.openNodeScope(jjtn001);
+            jjtn001.jjtSetFirstToken(getToken(0));
         try {
           jj_consume_token(LPAREN);
           switch ((jj_ntk==-1)?jj_ntk():jj_ntk) {
@@ -1150,6 +1202,7 @@
         } finally {
             if (jjtc001) {
               jjtree.closeNodeScope(jjtn001, true);
+              jjtn001.jjtSetLastToken(getToken(0));
             }
         }
         break;
@@ -1174,6 +1227,7 @@
     } finally {
           if (jjtc000) {
             jjtree.closeNodeScope(jjtn000, true);
+            jjtn000.jjtSetLastToken(getToken(0));
           }
     }
   }
@@ -1187,6 +1241,7 @@
   AstBracketSuffix jjtn000 = new AstBracketSuffix(JJTBRACKETSUFFIX);
   boolean jjtc000 = true;
   jjtree.openNodeScope(jjtn000);
+  jjtn000.jjtSetFirstToken(getToken(1));
     try {
       jj_consume_token(LBRACK);
       Expression();
@@ -1248,6 +1303,7 @@
         } finally {
             if (jjtc001) {
               jjtree.closeNodeScope(jjtn001, true);
+              jjtn001.jjtSetLastToken(getToken(0));
             }
         }
         break;
@@ -1272,6 +1328,7 @@
     } finally {
           if (jjtc000) {
             jjtree.closeNodeScope(jjtn000, true);
+            jjtn000.jjtSetLastToken(getToken(0));
           }
     }
   }
@@ -1313,15 +1370,18 @@
                                  /*@bgen(jjtree) Identifier */
                                   AstIdentifier jjtn000 = new AstIdentifier(JJTIDENTIFIER);
                                   boolean jjtc000 = true;
-                                  jjtree.openNodeScope(jjtn000);Token t = null;
+                                  jjtree.openNodeScope(jjtn000);
+                                  jjtn000.jjtSetFirstToken(getToken(1));Token t = null;
     try {
       t = jj_consume_token(IDENTIFIER);
                          jjtree.closeNodeScope(jjtn000, true);
                          jjtc000 = false;
+                         jjtn000.jjtSetLastToken(getToken(0));
                          jjtn000.setImage(t.image);
     } finally {
           if (jjtc000) {
             jjtree.closeNodeScope(jjtn000, true);
+            jjtn000.jjtSetLastToken(getToken(0));
           }
     }
   }
@@ -1334,7 +1394,8 @@
  /*@bgen(jjtree) Function */
         AstFunction jjtn000 = new AstFunction(JJTFUNCTION);
         boolean jjtc000 = true;
-        jjtree.openNodeScope(jjtn000);Token t0 = null;
+        jjtree.openNodeScope(jjtn000);
+        jjtn000.jjtSetFirstToken(getToken(1));Token t0 = null;
         Token t1 = null;
     try {
       t0 = jj_consume_token(IDENTIFIER);
@@ -1404,6 +1465,7 @@
     } finally {
           if (jjtc000) {
             jjtree.closeNodeScope(jjtn000, true);
+            jjtn000.jjtSetLastToken(getToken(0));
           }
     }
   }
@@ -1447,11 +1509,13 @@
           AstTrue jjtn001 = new AstTrue(JJTTRUE);
           boolean jjtc001 = true;
           jjtree.openNodeScope(jjtn001);
+          jjtn001.jjtSetFirstToken(getToken(1));
       try {
         jj_consume_token(TRUE);
       } finally {
           if (jjtc001) {
             jjtree.closeNodeScope(jjtn001, true);
+            jjtn001.jjtSetLastToken(getToken(0));
           }
       }
       break;
@@ -1459,11 +1523,13 @@
             AstFalse jjtn002 = new AstFalse(JJTFALSE);
             boolean jjtc002 = true;
             jjtree.openNodeScope(jjtn002);
+            jjtn002.jjtSetFirstToken(getToken(1));
       try {
         jj_consume_token(FALSE);
       } finally {
             if (jjtc002) {
               jjtree.closeNodeScope(jjtn002, true);
+              jjtn002.jjtSetLastToken(getToken(0));
             }
       }
       break;
@@ -1482,15 +1548,18 @@
                                        /*@bgen(jjtree) FloatingPoint */
                                         AstFloatingPoint jjtn000 = new AstFloatingPoint(JJTFLOATINGPOINT);
                                         boolean jjtc000 = true;
-                                        jjtree.openNodeScope(jjtn000);Token t = null;
+                                        jjtree.openNodeScope(jjtn000);
+                                        jjtn000.jjtSetFirstToken(getToken(1));Token t = null;
     try {
       t = jj_consume_token(FLOATING_POINT_LITERAL);
                                      jjtree.closeNodeScope(jjtn000, true);
                                      jjtc000 = false;
+                                     jjtn000.jjtSetLastToken(getToken(0));
                                      jjtn000.setImage(t.image);
     } finally {
           if (jjtc000) {
             jjtree.closeNodeScope(jjtn000, true);
+            jjtn000.jjtSetLastToken(getToken(0));
           }
     }
   }
@@ -1503,15 +1572,18 @@
                            /*@bgen(jjtree) Integer */
                             AstInteger jjtn000 = new AstInteger(JJTINTEGER);
                             boolean jjtc000 = true;
-                            jjtree.openNodeScope(jjtn000);Token t = null;
+                            jjtree.openNodeScope(jjtn000);
+                            jjtn000.jjtSetFirstToken(getToken(1));Token t = null;
     try {
       t = jj_consume_token(INTEGER_LITERAL);
                               jjtree.closeNodeScope(jjtn000, true);
                               jjtc000 = false;
+                              jjtn000.jjtSetLastToken(getToken(0));
                               jjtn000.setImage(t.image);
     } finally {
           if (jjtc000) {
             jjtree.closeNodeScope(jjtn000, true);
+            jjtn000.jjtSetLastToken(getToken(0));
           }
     }
   }
@@ -1524,15 +1596,18 @@
                          /*@bgen(jjtree) String */
                           AstString jjtn000 = new AstString(JJTSTRING);
                           boolean jjtc000 = true;
-                          jjtree.openNodeScope(jjtn000);Token t = null;
+                          jjtree.openNodeScope(jjtn000);
+                          jjtn000.jjtSetFirstToken(getToken(1));Token t = null;
     try {
       t = jj_consume_token(STRING_LITERAL);
                              jjtree.closeNodeScope(jjtn000, true);
                              jjtc000 = false;
+                             jjtn000.jjtSetLastToken(getToken(0));
                              jjtn000.setImage(t.image);
     } finally {
           if (jjtc000) {
             jjtree.closeNodeScope(jjtn000, true);
+            jjtn000.jjtSetLastToken(getToken(0));
           }
     }
   }
@@ -1546,11 +1621,13 @@
   AstNull jjtn000 = new AstNull(JJTNULL);
   boolean jjtc000 = true;
   jjtree.openNodeScope(jjtn000);
+  jjtn000.jjtSetFirstToken(getToken(1));
     try {
       jj_consume_token(NULL);
     } finally {
           if (jjtc000) {
             jjtree.closeNodeScope(jjtn000, true);
+            jjtn000.jjtSetLastToken(getToken(0));
           }
     }
   }
@@ -2064,6 +2141,21 @@
     return false;
   }
 
+//  private boolean jj_3R_73() {
+//    if (jj_scan_token(IDENTIFIER)) return true;
+//    return false;
+//  }
+//
+//  private boolean jj_3R_48() {
+//    Token xsp;
+//    xsp = jj_scanpos;
+//    if (jj_scan_token(51)) {
+//    jj_scanpos = xsp;
+//    if (jj_scan_token(52)) return true;
+//    }
+//    return false;
+//  }
+//
   /** Generated Token Manager. */
   public ELParserTokenManager token_source;
   SimpleCharStream jj_input_stream;
Index: src/main/java/com/sun/el/parser/Node.java
===================================================================
--- src/main/java/com/sun/el/parser/Node.java	(revision 352)
+++ src/main/java/com/sun/el/parser/Node.java	(working copy)
@@ -82,6 +82,14 @@
   
   public String getImage();
   
+  public Token jjtGetFirstToken();
+  public void jjtSetFirstToken(Token token);
+  public Token jjtGetLastToken();
+  public void jjtSetLastToken(Token token);
+
+  public int startOffset();
+  public int endOffset();
+
   public Object getValue(EvaluationContext ctx) throws ELException;
   public void setValue(EvaluationContext ctx, Object value) throws ELException;
   public Class getType(EvaluationContext ctx) throws ELException;
Index: src/main/java/com/sun/el/parser/SimpleCharStream.java
===================================================================
--- src/main/java/com/sun/el/parser/SimpleCharStream.java	(revision 352)
+++ src/main/java/com/sun/el/parser/SimpleCharStream.java	(working copy)
@@ -58,10 +58,13 @@
   public int bufpos = -1;
   protected int bufline[];
   protected int bufcolumn[];
+  private int bufoffset[];
 
   protected int column = 0;
   protected int line = 1;
 
+  private int offset = -1;
+
   protected boolean prevCharIsCR = false;
   protected boolean prevCharIsLF = false;
 
@@ -81,6 +84,7 @@
     char[] newbuffer = new char[bufsize + 2048];
     int newbufline[] = new int[bufsize + 2048];
     int newbufcolumn[] = new int[bufsize + 2048];
+    int newbufoffset[] = new int[bufsize + 2048];
 
     try
     {
@@ -98,6 +102,10 @@
         System.arraycopy(bufcolumn, 0, newbufcolumn, bufsize - tokenBegin, bufpos);
         bufcolumn = newbufcolumn;
 
+        System.arraycopy(bufoffset, tokenBegin, newbufoffset, 0, bufsize - tokenBegin);
+        System.arraycopy(bufoffset, 0, newbufoffset, bufsize - tokenBegin, bufpos);
+        bufoffset = newbufoffset;
+
         maxNextCharInd = (bufpos += (bufsize - tokenBegin));
       }
       else
@@ -111,6 +119,9 @@
         System.arraycopy(bufcolumn, tokenBegin, newbufcolumn, 0, bufsize - tokenBegin);
         bufcolumn = newbufcolumn;
 
+        System.arraycopy(bufoffset, tokenBegin, newbufoffset, 0, bufsize - tokenBegin);
+        bufoffset = newbufoffset;
+
         maxNextCharInd = (bufpos -= tokenBegin);
       }
     }
@@ -181,6 +192,7 @@
 
   protected void UpdateLineColumn(char c)
   {
+    offset++;
     column++;
 
     if (prevCharIsLF)
@@ -217,6 +229,7 @@
 
     bufline[bufpos] = line;
     bufcolumn[bufpos] = column;
+    bufoffset[bufpos] = offset;
   }
 
 /** Read a character. */
@@ -281,6 +294,10 @@
     return bufline[tokenBegin];
   }
 
+  public int offset() {
+      return bufoffset[tokenBegin];
+  }
+
 /** Backup a number of characters. */
   public void backup(int amount) {
 
@@ -296,11 +313,13 @@
     inputStream = dstream;
     line = startline;
     column = startcolumn - 1;
+    offset = -1;
 
     available = bufsize = buffersize;
     buffer = new char[buffersize];
     bufline = new int[buffersize];
     bufcolumn = new int[buffersize];
+    bufoffset = new int[buffersize];
   }
 
   /** Constructor. */
@@ -458,6 +477,7 @@
     buffer = null;
     bufline = null;
     bufcolumn = null;
+    bufoffset = null;
   }
 
   /**
Index: src/main/java/com/sun/el/parser/Token.java
===================================================================
--- src/main/java/com/sun/el/parser/Token.java	(revision 352)
+++ src/main/java/com/sun/el/parser/Token.java	(working copy)
@@ -48,6 +48,11 @@
 
 public class Token implements Serializable {
 
+   /**
+   * The offset for this token; needed by NetBeans.
+   */
+   public int offset;
+
   /**
    * An integer that describes the kind of this token.  This numbering
    * system is determined by JavaCCParser, and a table of these numbers is
Index: src/main/java/com/sun/el/parser/ELParser.jjt
===================================================================
--- src/main/java/com/sun/el/parser/ELParser.jjt	(revision 352)
+++ src/main/java/com/sun/el/parser/ELParser.jjt	(working copy)
@@ -47,6 +47,7 @@
 /* == Option Declaration == */
 options
 {
+        TRACK_TOKENS = true;
 	STATIC=false;
 	NODE_PREFIX="Ast";
 	VISITOR_EXCEPTION="javax.el.ELException";
@@ -65,12 +66,15 @@
 import javax.el.ELException;
 public class ELParser
 {
+
     public static Node parse(String ref) throws ELException
     {
         try {
-        	return (new ELParser(new StringReader(ref))).CompositeExpression();
+            ELParser parser = new ELParser(new StringReader(ref));
+            parser.token_source = new PatchedELParserTokenManager(parser.jj_input_stream);
+            return parser.CompositeExpression();
         } catch (ParseException pe) {
-           	throw new ELException(pe.getMessage());
+                throw new ELException(pe.getMessage());
         }
     }
 }
Index: build.xml
===================================================================
--- build.xml	(revision 352)
+++ build.xml	(working copy)
@@ -44,7 +44,7 @@
 
 <project name="el-impl" default="generate" basedir=".">
 
-    <property name="javacc.home" value="/home/kchung/tools/javacc-5.0"/>
+    <property name="javacc.home" value="/home/marfous/Programs/Utils/javacc-5.0"/>
     <property name="dir" value="src/main/java/com/sun/el/parser"/>
 
     <target name="generate" description="Generate java files">
