Index: src/main/java/org/apache/lucene/index/ExtendedIndexWriter.java
===================================================================
--- src/main/java/org/apache/lucene/index/ExtendedIndexWriter.java	(revision 7197)
+++ src/main/java/org/apache/lucene/index/ExtendedIndexWriter.java	(working copy)
@@ -1,34 +0,0 @@
-package org.apache.lucene.index;
-
-import java.io.IOException;
-
-import org.apache.lucene.analysis.Analyzer;
-import org.apache.lucene.store.Directory;
-import org.apache.lucene.store.LockObtainFailedException;
-
-/**
- * Hacky class.
- * 
- * @author cstamas
- */
-public class ExtendedIndexWriter
-    extends IndexWriter
-{
-    @Deprecated
-    public ExtendedIndexWriter( Directory d, boolean autoCommit, Analyzer a, boolean create )
-        throws CorruptIndexException, LockObtainFailedException, IOException
-    {
-        super( d, autoCommit, a, create );
-    }
-
-    public ExtendedIndexWriter( Directory d, Analyzer a, boolean create, MaxFieldLength mfl )
-        throws CorruptIndexException, LockObtainFailedException, IOException
-    {
-        super( d, a, create, mfl );
-    }
-
-    public boolean hasUncommitedChanges()
-    {
-        return pendingCommit != null;
-    }
-}
Index: src/main/java/org/sonatype/nexus/index/context/DefaultIndexingContext.java
===================================================================
--- src/main/java/org/sonatype/nexus/index/context/DefaultIndexingContext.java	(revision 7197)
+++ src/main/java/org/sonatype/nexus/index/context/DefaultIndexingContext.java	(working copy)
@@ -431,7 +431,7 @@
             }
             else
             {
-                return indexWriter.hasUncommitedChanges();
+                return indexWriter.hasUncommittedChanges();
             }
         }
     }
Index: src/main/java/org/sonatype/nexus/index/context/NexusIndexWriter.java
===================================================================
--- src/main/java/org/sonatype/nexus/index/context/NexusIndexWriter.java	(revision 7197)
+++ src/main/java/org/sonatype/nexus/index/context/NexusIndexWriter.java	(working copy)
@@ -6,10 +6,13 @@
 package org.sonatype.nexus.index.context;
 
 import java.io.IOException;
+import java.lang.reflect.Field;
+import java.util.logging.Level;
+import java.util.logging.Logger;
 
 import org.apache.lucene.analysis.Analyzer;
 import org.apache.lucene.index.CorruptIndexException;
-import org.apache.lucene.index.ExtendedIndexWriter;
+import org.apache.lucene.index.IndexWriter;
 import org.apache.lucene.store.Directory;
 import org.apache.lucene.store.LockObtainFailedException;
 
@@ -19,7 +22,7 @@
  * to allow to track if writer is closed
  */
 public class NexusIndexWriter
-    extends ExtendedIndexWriter
+    extends IndexWriter
 {
     private boolean closed;
 
@@ -50,4 +53,20 @@
     {
         return closed;
     }
+
+    boolean hasUncommittedChanges()
+    {
+        try
+        {
+            Field pendingCommit = IndexWriter.class.getDeclaredField( "pendingCommit" );
+            pendingCommit.setAccessible( true );
+            return pendingCommit.get( this ) != null;
+        }
+        catch ( Exception x )
+        {
+            Logger.getLogger( NexusIndexWriter.class.getName() ).log( Level.WARNING, null, x );
+            return false;
+        }
+    }
+
 }
