diff --git a/src/main/org/jvyamlb/ScannerImpl.java b/src/main/org/jvyamlb/ScannerImpl.java
index 5f2d66c..ea680e9 100644
--- a/src/main/org/jvyamlb/ScannerImpl.java
+++ b/src/main/org/jvyamlb/ScannerImpl.java
@@ -859,6 +859,7 @@
                 chunks.append(this.buffer.bytes,this.pointer,length);
                 length = 0;
                 chunks.append(scanUriEscapes(name));
+                break;
             } else {
                 length++;
             }
