--- a/web.css/src/org/netbeans/modules/css/parser/CSSParserTokenManager.java  Wed Feb 13 13:07:58 2008 +0100
+++ b/web.css/src/org/netbeans/modules/css/parser/CSSParserTokenManager.java  Wed Feb 13 13:24:52 2008 +0100
@@ -3358,8 +3358,7 @@ protected Token jjFillToken()
    endColumn = input_stream.getEndColumn();
    t = Token.newToken(jjmatchedKind, tokenImage);

+   t.offset = input_stream.offset();
    
    t.beginLine = beginLine;
    t.endLine = endLine;
